pkg/proctree/tree.go

```diff
@@ -0,0 +1,297 @@
+	MaxProcesses int
+	// Max amount of threads nodes to allow in memory.
+	// Too small value here might result missing information in the tree and inconsistency.
+	// Default recommended value is 32768, as this is the max amount of TIDs in the system.
+	// You might even want more than 32768 to allow relevant information of dead threads to be
+	// available after their exit.
+	MaxThreads int
+	// Max size of cache for processes removing after exit - translate to the delay between process
+	// exit to its removal from the tree (if it is not interesting after its death).
+	// Too small value here might result processes missing from tree after their exit, before
+	// users were able to query the process information because of the delay between the feeding
+	// of the tree and the consumption of its information.
+	MaxCacheDelete int
+}
+
+// ProcessTree is a struct which follow the state of processes during runtime of a system.
+// The process tree is updated through Tracee's events, and is designed to overcome problems that
+// may arise because of events consumption (like handling lost events).
+// For more information on the logic, please go to the package documentation.
+type ProcessTree struct {
+	processes          types.RWMap[int, *processNode]
+	threads            types.RWMap[int, *threadNode]
+	processesGC        *lru.Cache[nodeUniqueId, *processNode]
+	threadsGC          *lru.Cache[nodeUniqueId, *threadNode]
+	deadProcessesCache *lru.Cache[int, bool]
+}
+
+func NewProcessTree(config ProcessTreeConfig) (*ProcessTree, error) {
+	tree := &ProcessTree{
+		processes: types.InitRWMap[int, *processNode](),
+		threads:   types.InitRWMap[int, *threadNode](),
+	}
+	cache, err := lru.NewWithEvict[int, bool](
+		config.MaxCacheDelete,
+		func(dpid int, _ bool) {
+			tree.removeProcessFromTree(dpid)
+		})
+	if err != nil {
+		return nil, err
+	}
+	processesGC, err := lru.NewWithEvict[nodeUniqueId, *processNode](
+		config.MaxProcesses,
+		func(id nodeUniqueId, node *processNode) {
+			err = tree.deleteProcess(node)
+			if err != nil {
+				logger.Errorw("delete process from tree", "err", err.Error())
+			}
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	threadsGC, err := lru.NewWithEvict[nodeUniqueId, *threadNode](
+		config.MaxThreads,
+		func(id nodeUniqueId, node *threadNode) {
+			err = tree.deleteThread(node)
+			if err != nil {
+				logger.Errorw("delete thread from tree", "err", err.Error())
+			}
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	tree.deadProcessesCache = cache
+	tree.processesGC = processesGC
+	tree.threadsGC = threadsGC
+	return tree, nil
+}
+
+// getProcess get the process node from the process tree if exists, and return error if not.
+func (tree *ProcessTree) getProcess(hostProcessID int) (*processNode, error) {
+	process, ok := tree.processes.Get(hostProcessID)
+	if !ok {
+		return nil, fmt.Errorf("no process with given Id is recorded")
+	}
+	_, _ = tree.processesGC.Get(process.GetUniqueId())
```

Comment:
// keep cache hot ?

Reply:
I want here to update its place in the LRU.
I will add a comment on it.